### ISSUE TITLE
Problems building memkind trunk 

### ISSUE THREAD

**Issue by adrianjhpc** [memkind/memkind#711]
I'm trying to build the memkind trunk but I get this error:

```
[ ! -e jemalloc/configure ] && (cd jemalloc && autoconf) || exit 0
[ ! -e jemalloc/lib/libjemalloc_pic.a ] && (cd jemalloc && ./configure --enable-autogen  --without-export --with-version=5.2.1-0-gea6b3e973b477b8061e0076bb257dbd7f3faa756 --disable-fill  --disable-initial-exec-tls --with-jemalloc-prefix=jemk_ --with-malloc-conf="narenas:256,lg_tcache_max:12"  && make) || exit 0
  CC       src/heap_manager.lo
src/heap_manager.c(32): error: identifier "jemk_malloc_usable_size" is undefined
      .heap_manager_malloc_usable_size = jemk_malloc_usable_size,
```

I've tried with both the Intel and GNU compilers but get the same results.

This build is configured without any specific options.

**Comment by bratpiorka** [comment]
@adrianjhpc Do you still have problems with building Memkind? Or can I close this issue?

**Comment by adrianjhpc** [comment]
I'm still having the same issue.

This is a CentOS 7.9 system, and I'm building with the Intel OneAPI compilers. If I follow your recipe above I get these errors:

```
[ ! -e jemalloc/configure ] && (cd jemalloc && autoconf) || exit 0
[ ! -e jemalloc/lib/libjemalloc_pic.a ] && (cd jemalloc && ./configure --enable-autogen  --without-export --with-version=5.2.1-0-gea6b3e973b477b8061e0076bb257dbd7f3faa756 --disable-fill  --disable-initial-exec-tls --with-jemalloc-prefix=jemk_ --with-malloc-conf="narenas:256,lg_tcache_max:12"  && make) || exit 0
  CC       src/heap_manager.lo
src/heap_manager.c(32): error #144: a value of type "size_t={unsigned long} (*)(const void *)" cannot be used to initialize an entity of type "size_t={unsigned long} (*)(void *)"
      .heap_manager_malloc_usable_size = jemk_malloc_usable_size,
                                         ^

compilation aborted for src/heap_manager.c (code 2)
make: *** [src/heap_manager.lo] Error 1
```

I can't see a version of jemalloc anywhere on the system, but I could be missing something.

Do you want more details?

**Comment by bratpiorka** [comment]
Adrian, sorry for the very long delay, but I only now have time to address your problem. Is it still accutable? I would like to start by reproducing it

**Comment by adrianjhpc** [comment]
No worries, I've not updated the trunk and checked recently, I'll try that.

**Comment by adrianjhpc** [comment]
There didn't seem to be any updates from the trunk (I'm not sure that is correct), so it is still occuring for me.

This is with:
```
               Memkind version 1.12.0+dev13+g4b86f93
```

**Comment by bratpiorka** [comment]
How do you compile memkind? you set CC to icc? icx? what about configure and make flags? what is the version of Intel compiler?

**Comment by adrianjhpc** [comment]
Initially all I was configuring with was a prefix specifier, and that was picking up the default system compiler (gcc 4.8.5).

I've also tried with the icc version 2021.3.0 compiler, again with no other configuration options specified. That gives this error on compile:

```
[ ! -e jemalloc/configure ] && (cd jemalloc && autoconf) || exit 0
[ ! -e jemalloc/lib/libjemalloc_pic.a ] && (cd jemalloc && ./configure --enable-autogen  --without-export --with-version=5.2.1-0-gea6b3e973b477b8061e0076bb257dbd7f3faa756 --disable-fill  --disable-initial-exec-tls --with-jemalloc-prefix=jemk_ --with-malloc-conf="narenas:256,lg_tcache_max:12"  && make) || exit 0
  CC       src/heap_manager.lo
src/heap_manager.c(32): error #144: a value of type "size_t={unsigned long} (*)(const void *)" cannot be used to initialize an entity of type "size_t={unsigned long} (*)(void *)"
      .heap_manager_malloc_usable_size = jemk_malloc_usable_size,
                                         ^

compilation aborted for src/heap_manager.c (code 2)
make: *** [src/heap_manager.lo] Error 1
```

**Comment by bratpiorka** [comment]
OK I see where is the problem - during building, jemalloc tests whether malloc_usable_size definition can use const argument or not and defines je_malloc_usable_size() base on that. Memkind doesn't do this (it always uses non-const version). I will add similar configuration code to Memkind to make this definition common.

**Comment by bratpiorka** [comment]
Could you please try if https://github.com/memkind/memkind/pull/713/commits fixes this problem?

**Comment by adrianjhpc** [comment]
Yes, looks like this fixes that problem. It throws up a couple more:

- No `threads.h` on the system (required by `tiering/memtier.c`)
- Unhappy about the `\` in this line `int ret = sscanf(query, "%[^\[]%n", name_substr, &chr_read);` in  `src/memkind_memtier.c`  (I think it might need double escaped).

But I think this can be closed now.

**Comment by bratpiorka** [comment]
closed by #713
